### ISSUE TITLE
Impement `@precomupte` decorator

### ISSUE THREAD

**Issue by saidctb** [pyccel/pyccel#1818]
Previously, the `@python decorator was utilized to signal to the compiler that a function should be executed at compile time, and its result should be directly incorporated into the code. However, this mechanism was removed in the issue https://github.com/pyccel/pyccel/issues/668 because it was not handled properly.
I want to reintroduce this mechanism and make it more general than before as I think it can be useful in many instances, where we need to precompute some expressions and hardcode them directly in the code to make the code faster.
Additionally, it can be used in unit testing and generate an executable that does everything instead of separating the implementation from the unit testing.
Key Features:
- The function that contains the `@precompute` decorator should be able to call other functions in the same file or access the imported files. previously the function with the `@python`  didn't have access to other functions in the same file.
- The results from calling a function with the `@precompute` decorator should be declared as `Final`, to ensure immutability.
- The translated code will not contain those functions unless we can handle callbacks. In that case, Fortran/C can call the Python implementation of those functions.

Here I constructed many examples to illustrate its usefulness.
- Compute matrix-matrix product and use `precompute` decorator to generate test cases:
```python
import numpy as np

# Matrix Matrix product
def matmat(A:'float[:,:]', B:'float[:,:]'):
    """
    Compute the matrix matrix product.

    """
    ...

    return C
#======================================================
#UNIT TESTING
# Use the precompute decorator to create the test cases in the compilation time
@precompute
def generate_matrices(n:'int'):
    """ Generate matrices and compute their product using the numpy library"""
    A = np.random.rand(n,n)
    B = np.random.rand(n,n)
    C = A@B
    return A,B,C

if __name__ == '__main__':

    A1,B1,C1 = generate_matrices(5)
    A2,B2,C2 = generate_matrices(8)
    A3,B3,C3 = generate_matrices(12)

    err1 = abs(C1-matmat(A1,B1)).sum()
    err2 = abs(C2-matmat(A2,B2)).sum()
    err3 = abs(C3-matmat(A3,B3)).sum()
    for r in (err1, err2, err3):
        if abs(r)<tol:
            print("PASSED!")
        else:
            print('FAILED!')
```
- Compute numerical integrals and check the results using the Sympy library:
```python
import numpy as np

# Define the functions to be integrated
def f1(x:'float'):
    return x**2 + np.sin(x)

def f2(x:'float'):
    return np.cos(x)
    
def f3(x:'float'):
    return np.tan(x)**2
    
def f4(x:'float'):
    return x**2 + x**3

# numerical integration
def numerical_integral(func:'(float)(float)', a:'float', b:'float'):
    """
    Compute the definite integral of a function numerically using the trapezoidal rule.
    """
    ...
    return result

#======================================================
#UNIT TESTING
def numerical_integrals(a:'float',b:'float'):
    integral1 = numerical_integral(f1, a, b)
    integral2 = numerical_integral(f2, a, b)
    integral3 = numerical_integral(f3, a, b)
    integral4 = numerical_integral(f4, a, b)
    return integral1,integral2, integral3, integral4

# Use the precompute decorator to compute exact integrals
@precompute
def exact_integrals(a:'float',b:'float'):
    from sympy.abc import x
    integral1 = sp.integrate(f1(x), (x, a, b))
    integral2 = sp.integrate(f2(x), (x, a, b))
    integral3 = sp.integrate(f3(x), (x, a, b))
    integral4 = sp.integrate(f4(x), (x, a, b))
    return integral1,integral2, integral3, integral4

if __name__ == '__main__':

    num = numerical_integrals(0.,1.)
    ext = exact_integrals(0.,1.)
    for r1,r2 in zip(num, ext):
        if abs(r1-r2)<tol:
            print("PASSED!")
        else:
            print('FAILED!')
````
- Compute the minimum of a function using gradient descent algorithm and compare with an exact solution:
```python
import numpy as np

# Define the function to be optimized
def f(x:'float'):
    return (x - 2) ** 2 + np.sin(x)

# Gradient descent optimization
def gradient_descent(func, initial_guess, learning_rate=0.01, max_iter=1000, tol=1e-6):
    x = initial_guess
    for _ in range(max_iter):
        grad = (func(x + tol) - func(x)) / tol  # Approximate gradient
        x_new = x - learning_rate * grad
        if abs(x_new - x) < tol:
            break
        x = x_new
    return x

#======================================================
# UNIT TESTING
@precompute
def symbolic_optimization():
    import sympy as sp
    x = sp.Symbol('x')
    gradient = sp.diff(f(x), x)
    critical_points = sp.solve(gradient, x)
    return float(critical_points)

@precompute
def exact_critical_points():
    return symbolic_optimization()

if __name__ == '__main__':
    initial_guess = 0  # Initial guess for numerical optimization

    # Numerical optimization using gradient descent
    numerical_result = gradient_descent(f, initial_guess)

    # Exact optimization
    exact_result = exact_critical_points()

    # Unit test
    if np.isclose(numerical_result, exact_result):
        print("PASSED!")
    else:
        print("FAILED!")
````

**Comment by EmilyBourne** [comment]
What is the difference between this ` @precompute` decorator and the `@sympy` decorator?
